map_i80.py

```diff
@@ -483,76 +483,85 @@ def step(self, policy_action=None):
-            draw_line(s, w, (m + 60 * LANE_W, bottom + 3), (2 * m + sw, bottom), 1)
-
-            # offroad regions
-            pygame.Surface.fill(s, b, pygame.Rect(m + 0, m + lanes[0]['min']-35, sw, 34))
-            pygame.draw.polygon(s, b, [
-                (m + 0, bottom+2),
-                (m + 0, bottom + 29-1),
-                (m + 18 * LANE_W, bottom + 29-1 - slope * 18 * LANE_W),
-                (m + 18 * LANE_W, bottom+2)
-            ])
-            pygame.draw.polygon(s, b, [
-                (m + 0, bottom + 54),
-                (m + 0, bottom + 54+30),
-                (m + 60 * LANE_W, bottom + 54+30),
-                (m + 60 * LANE_W, bottom + 54 - slope * 60 * LANE_W)
-            ])
-            pygame.Surface.fill(
-                s,
-                b,
-                pygame.Rect(m + 60 * LANE_W, bottom + 5, sw-60*LANE_W, 54+30-5))
-
-            self._lane_surfaces[mode] = surface.copy()
+            if colored_lane is None:
+                s = surface  # screen
+                draw_line = pygame.draw.line  # shortcut
+                w = colours['r']  # colour white
+                b = colours['b']  # colour blue
+                sw = self.screen_size[0]  # screen width
+                m = offset
+
+                for lane in lanes:
+                    draw_line(s, w, (0, lane['min'] + m), (sw + 2 * m, lane['min'] + m), 1)
+
+                bottom = lanes[-1]['max'] + m
+                draw_line(s, w, (0, bottom), (m + 18 * LANE_W, bottom), 1)
+                draw_line(s, w, (m, bottom + 29), (m + 18 * LANE_W, bottom + 29 - slope * 18 * LANE_W), 1)
+                draw_line(s, w, (m + 18 * LANE_W, bottom + 13), (m + 31 * LANE_W, bottom), 1)
+                draw_line(s, w, (m, bottom + 53), (m + 60 * LANE_W, bottom + 53 - slope * 60 * LANE_W), 1)
+                draw_line(s, w, (m + 60 * LANE_W, bottom + 3), (2 * m + sw, bottom), 1)
+
+                # offroad regions
+                pygame.Surface.fill(s, b, pygame.Rect(m + 0, m + lanes[0]['min'] - 35, sw, 34))
+                pygame.draw.polygon(s, b, [
+                    (m + 0, bottom + 2),
+                    (m + 0, bottom + 29 - 1),
+                    (m + 18 * LANE_W, bottom + 29 - 1 - slope * 18 * LANE_W),
+                    (m + 18 * LANE_W, bottom + 2)
+                ])
+                pygame.draw.polygon(s, b, [
+                    (m + 0, bottom + 54),
+                    (m + 0, bottom + 54 + 30),
+                    (m + 60 * LANE_W, bottom + 54 + 30),
+                    (m + 60 * LANE_W, bottom + 54 - slope * 60 * LANE_W)
+                ])
+                pygame.Surface.fill(
+                    s,
+                    b,
+                    pygame.Rect(m + 60 * LANE_W, bottom + 5, sw - 60 * LANE_W, 54 + 30 - 5))
+
+                self._lane_surfaces[mode] = surface.copy()
+            else:
+                s = surface  # screen
+                lane = pygame.image.load(colored_lane)
```

Comment:
This is wrong.
The `'machine'` view has an `m` offset. Did you try saving the `Surface` you're getting?

Reply:
![i80-machine](https://user-images.githubusercontent.com/29690116/85820863-3a7b3180-b745-11ea-9926-dc44d50759c5.png)
![i80-machine-1](https://user-images.githubusercontent.com/29690116/85820867-3c44f500-b745-11ea-9048-6cea6b41d1e0.png)

I add offset for machine mode. Seems works well. Should I add blue background for colored map?

Reply:
I think not. The rails are going to be attractors. Before, the red and blue were repellers.